lib/framework/StatefulService.h

```diff
@@ -60,66 +71,50 @@ class StatefulService {
     }
   }
 
-  void updateWithoutPropagation(std::function<void(T&)> callback) {
-#ifdef ESP32
-    xSemaphoreTakeRecursive(_accessMutex, portMAX_DELAY);
-#endif
-    callback(_state);
-#ifdef ESP32
-    xSemaphoreGiveRecursive(_accessMutex);
-#endif
+  StateUpdateResult update(std::function<StateUpdateResult(T&)> stateUpdater, const String& originId) {
+    beginTransaction();
+    StateUpdateResult result = stateUpdater(_state);
+    if (result == StateUpdateResult::CHANGED) {
+      callUpdateHandlers(originId);
+    }
+    endTransaction();
+    return result;
   }
 
-  void updateWithoutPropagation(JsonObject& jsonObject, JsonDeserializer<T> deserializer) {
-#ifdef ESP32
-    xSemaphoreTakeRecursive(_accessMutex, portMAX_DELAY);
-#endif
-    deserializer(jsonObject, _state);
-#ifdef ESP32
-    xSemaphoreGiveRecursive(_accessMutex);
-#endif
+  StateUpdateResult updateWithoutPropagation(std::function<StateUpdateResult(T&)> stateUpdater) {
+    beginTransaction();
+    StateUpdateResult result = stateUpdater(_state);
+    endTransaction();
+    return result;
   }
 
-  void update(std::function<void(T&)> callback, const String& originId) {
-#ifdef ESP32
-    xSemaphoreTakeRecursive(_accessMutex, portMAX_DELAY);
-#endif
-    callback(_state);
-    callUpdateHandlers(originId);
-#ifdef ESP32
-    xSemaphoreGiveRecursive(_accessMutex);
-#endif
+  StateUpdateResult update(JsonObject& jsonObject, JsonStateUpdater<T> stateUpdater, const String& originId) {
+    beginTransaction();
+    StateUpdateResult result = stateUpdater(jsonObject, _state);
+    if (result == StateUpdateResult::CHANGED) {
+      callUpdateHandlers(originId);
+    }
+    endTransaction();
+    return result;
```

Comment:
```suggestion
    beginTransaction();
    StateUpdateResult result = stateUpdater(jsonObject, _state);
    endTransaction();
    if (result == StateUpdateResult::CHANGED) {
      callUpdateHandlers(originId);
    }
    return result;
```

We can close transaction much earlier, right after state changed.